### PR TITLE
[WIP] Add new scorer: MaNoScorer

### PR DESCRIPTION
**Pull request by ambroiseodt**
Implementation of [MaNo](https://arxiv.org/pdf/2405.18979) as an unsupervised realistic scorer.

This should answer #288.
 
- [x] Update README.md
- [x] Update docs/source/all.rst
- [x] Update skada/metrics.py
- [x] Update skada/tests/test_scorer.py
- [x] Update skada/deep/tests/test_deep_scorer.py

### CODE REVIEW

**Comment by antoinecollas** [comment]
Yes, deep learning tests are failling because skorch v1.0.0 is not compatible with sklearn 1.6.0. They fixed skorch and we are waiting for a new release (see https://github.com/skorch-dev/skorch/issues/1085).

You should still be able to test your PR with skorch 1.0.0 and sklearn 1.5.2.

**Comment by ambroiseodt** [comment]
Thanks @antoinecollas, yes I am using skorch 1.0.0 and sklearn 1.5.2 in local, I will push the test_deep_scorers once it works locally for me

**Comment by ambroiseodt** [comment]
> Thanks @antoinecollas, yes I am using skorch 1.0.0 and sklearn 1.5.2 in local, I will push the test_deep_scorers once it works locally for me

@antoinecollas It seems that the skorch issue has been resolved :)

File: skada/metrics.py [hunk 1]
```diff
@@ -782,3 +784,131 @@ def _score(self, estimator, X, y=None, sample_domain=None, **params):
...
+        logits -= np.max(logits, axis=1, keepdims=True)
+        exp_logits = np.exp(logits)
+        exp_logits /= np.sum(exp_logits, axis=1, keepdims=True)
+        return exp_logits
+
+    @staticmethod
+    def _taylor_softmax(logits):
+        """Compute Taylor approximation of order 2 of softmax."""
+        tay_logits = 1 + logits + logits**2 / 2
+        tay_logits -= np.min(tay_logits, axis=1, keepdims=True)
```
Comment (tgnassou) [inline skada/metrics.py:914]: Where does the min come from in the eq ? I didn't find it in the paper

File: skada/tests/test_scorer.py [hunk 2]
```diff
@@ -351,6 +355,44 @@ def test_mixval_scorer_regression(da_reg_dataset):
...
+    estimator.fit(X, y, sample_domain=sample_domain)
+
+    scorer = MaNoScorer()
+    score_mean = scorer._score(estimator, X, y, sample_domain=sample_domain)
+    assert isinstance(score_mean, float), "score_mean is not a float"
+
+    # Test softmax normalization
+    scorer = MaNoScorer(threshold=-1)
+    score_mean = scorer._score(estimator, X, y, sample_domain=sample_domain)
+    assert isinstance(score_mean, float), "score_mean is not a float"
```
Comment (tgnassou) [inline skada/tests/test_scorer.py:374]: You don't check if it is the right normalization which is considered no?

File: skada/deep/tests/test_deep_scorer.py [hunk 3]
```diff
@@ -191,7 +196,14 @@ def test_dev_scorer_on_source_only(da_dataset):
...
 
-def test_dev_exception_layer_name(da_dataset):
+@pytest.mark.parametrize(
+    "scorer",
+    [
+        DeepEmbeddedValidation(),
+        MaNoScorer(),
+    ],
+)
+def test_dev_exception_layer_name(scorer, da_dataset):
```
Comment (tgnassou) [inline skada/deep/tests/test_deep_scorer.py:206]: change name

Comment (ambroiseodt) [reply skada/deep/tests/test_deep_scorer.py:206]: To what? it was the name already given

File: skada/metrics.py (hunk 1 above)
Comment (ambroiseodt) [reply skada/metrics.py:914]: It is not needed when p=2 but necessary for p>=3 to ensure that the outputs is a probability. It is explained in footnote 2 of page 6 in https://arxiv.org/pdf/2405.18979

File: skada/deep/tests/test_deep_scorer.py (hunk 3 above)
Comment (ambroiseodt) [reply skada/deep/tests/test_deep_scorer.py:206]: understood, will do

File: skada/tests/test_scorer.py (hunk 2 above)
Comment (ambroiseodt) [reply skada/tests/test_scorer.py:374]: Understood, will do

**Review by tgnassou — APPROVED** [review]
